chia/wallet/did_wallet/did_wallet_puzzles.py

```diff
@@ -82,24 +74,68 @@ def get_innerpuzzle_from_puzzle(puzzle: Program) -> Optional[Program]:
     return INNER_PUZZLE
 
 
-def create_recovery_message_puzzle(recovering_coin_id: bytes32, newpuz: bytes32, pubkey: G1Element):
+def create_recovery_message_puzzle(recovering_coin_id: bytes32, newpuz: bytes32, pubkey: G1Element) -> Program:
     puzstring = f"(q . ((0x{ConditionOpcode.CREATE_COIN_ANNOUNCEMENT.hex()} 0x{recovering_coin_id.hex()}) (0x{ConditionOpcode.AGG_SIG_UNSAFE.hex()} 0x{bytes(pubkey).hex()} 0x{newpuz.hex()})))"  # noqa
-    puz = binutils.assemble(puzstring)
+    puz = assemble(puzstring)
     return Program.to(puz)
 
 
-def create_spend_for_message(parent_of_message, recovering_coin, newpuz, pubkey):
+def create_spend_for_message(
+    parent_of_message: bytes32, recovering_coin: bytes32, newpuz: bytes32, pubkey: G1Element
+) -> CoinSpend:
     puzzle = create_recovery_message_puzzle(recovering_coin, newpuz, pubkey)
     coin = Coin(parent_of_message, puzzle.get_tree_hash(), uint64(0))
     solution = Program.to([])
     coinsol = CoinSpend(coin, puzzle, solution)
     return coinsol
 
 
+def match_did_puzzle(puzzle: Program) -> Tuple[bool, Iterator[Program]]:
+    """
+    Given a puzzle test if it's an DID and, if it is, return the curried arguments
+    """
+    try:
+        mod, curried_args = puzzle.uncurry()
+        if mod == SINGLETON_TOP_LAYER_MOD:
+            mod, curried_args = curried_args.rest().first().uncurry()
+            if mod == DID_INNERPUZ_MOD:
+                return True, curried_args.as_iter()
+    except Exception:
+        import traceback
+
+        print(f"exception: {traceback.format_exc()}")
+        return False, iter(())
+    return False, iter(())
+
+
 # inspect puzzle and check it is a DID puzzle
-def check_is_did_puzzle(puzzle: Program):
+def check_is_did_puzzle(puzzle: Program) -> bool:
     r = puzzle.uncurry()
     if r is None:
         return r
     inner_f, args = r
     return is_did_core(inner_f)
+
+
+def metadata_to_program(metadata: Dict) -> Program:
+    """
+    Convert the metadata dict to a Chialisp program
+    :param metadata: User defined metadata
+    :return: Chialisp program
+    """
+    kv_list = []
+    for key, value in metadata.items():
+        kv_list.append((assemble(key), assemble(value)))
+    return Program.to(kv_list)
+
+
+def program_to_metadata(program: Program) -> Dict:
+    """
+    Convert a program to a metadata dict
+    :param program: Chialisp program contains the metadata
+    :return: Metadata dict
+    """
+    metadata = {}
+    for key, val in program.as_iter():
+        metadata[str(key.as_python(), "ascii")] = str(val.as_python(), "ascii")
```

Comment:
Is the ascii really the right encoding to use? I would have thought that the default of UTF-8 would be more appropriate.

Reply:
It depends if we want to support utf-8 chars. Anyway, I changed it and covered this in the test.